docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,210 @@
+
+#### Option 2: New API to Match Tombstones
+
+Since the block indexes are already downloaded by the store gateway, it can already scan the blocks and calculate the tombstones. So a way here is to expose some APIs from store gateway, and then compactor can query the APIs to know what blocks can be deleted.
+
+The gRPC API is called `Tombstone` API. The compactor can either query from the store gateway directly or query from queriers, and queriers do the API federation from multiple store gateway instances.
+
+The example gRPC format looks like below. The response is almost the same as the `MemTombstone` result.
+
+```protobuf
+service Tombstone {
+  rpc Tombstone(TombstoneRequest) returns (TombstoneResponse);
+}
+
+message TombstoneRequest {
+  int64 min_time = 1;
+  int64 max_time = 2;
+  repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];
+}
+
+message TombstoneResponse {
+  repeated SeriesIntervals series_intervals = 1;
+}
+
+message SeriesIntervals {
+  int64 series_ref = 1;
+  repeated Interval intervals = 2;
+}
+
+message Interval {
+  int64 start = 1;
+  int64 end = 2;
+}
+```
+
+After getting the query result, the compactor does almost the same thing as the previous option: generate the tombstone file during compaction time.
+
+The drawback for this approach is that we are inventing a new API for this purpose only, which might be a bit of pain in the future.
+
+#### Option 3: Delete Series using Rewrite Logic
+
+The previous two options do the series deletion at compaction time, this option does actual series deletion using block rewrite. For the index scanning, it is still required to choose either option 1 or 2 to avoid unncessary rewrite even if the block doesn't have any matched series.
+
+This is not so efficient as rewrite is a different process than compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.
+
+#### Considerations
+
+Overall, among the 3 options, option 1 is better as it is easier to implement and straightforward. No additional API required or no additional logic change on the compaction side.
+
+### Other Design Considerations
+
+#### Compaction Planner for Tombstone
+
+If series deletion is done at compaction time, then no much thing to change at the planner as we just need to delete series at compaction time. For the ratio calculation, as the tombstone results are pre-calculated it is easy to know the ratio as well. The threshold can be made configurable too.
+
+The exception is for blocks that at max compaction level, no compaction will happen to it and the tombstones will remain forever if the ratio is lower than the deletion threshold. In this case the planner can check the block compaction level first, if it is max compaction level and the tombstone exists for some period of time, then do compaction for that block.
```

Comment:
Is the ratio going to be calculated per each tombstone or based on the total of all tombstones that can act on a particular block? Asking because even if one tombstone doesn't reach above the threshold, the sum of multiple smaller tombstones might.

Reply:
It is for all tombstones. We calculate the number of series to be deleted for all tombstones that match this block.